theano/compile/function_module.py

```diff
@@ -483,6 +488,66 @@ def __copy__(self):
                 there.data = here.data
         return cpy
 
+    def copy(self, share_memory = False):
+        """
+        Copy this function. Copied function will have separated maker and fgraph
+        with original function. User can choose whether to separate storage by 
+        changing the share_memory arguments
+        ---------------------
+        Params:
+            share_memory -- { boolean } Default is False. When True, two function
+            share intermediate storages(storages except input and output storages)
+        ---------------------
+        Returns:
+            func -- Copied theano.Function
+        """
+
+        if not share_memory:
+            return self.__copy__()
+        else:
+            # copy SymbolocKits
+            ins, outs = copy.deepcopy([self.maker.inputs, self.maker.outputs])
```

Comment:
Many thanks :)

I used to believe that the value of SharedVariable was updated by storage_map and ```In()```s, so I was struggling about this. Now I know it's not the case. I got the idea of how this should be done properly( I hope ) as I'll update the code tomorrow morning.

And, I'll take notes about possible improvement from now on.